### PR TITLE
Read auth_tries from config instead using 3

### LINKED ISSUE #59174: [BUG] The setting of auth_tries is lost and set to 3 when going offline
**Description**

The setting of `auth_tries` is lost and set to 3 when going offline

**Setup**
Write into your minion config
```
auth_timeout: 60
auth_tries: 60
ping_interval: 10
```

**Steps to Reproduce the behavior**

Stop the minion

Go Offline

Start the minion

Find in the log:
- auth_timeout: 60
- auth_tries: 60

```
2020-12-19 13:44:49,574 [salt.minion      :1089][ERROR   ][12596] Minion unable to successfully connect to a Salt Master.
2020-12-19 13:44:53,224 [salt.transport.zeromq:1363][DEBUG   ][12596] SaltReqTimeoutError, retrying. (1/60)
2020-12-19 13:45:53,226 [salt.transport.zeromq:1363][DEBUG   ][12596] SaltReqTimeoutError, retrying. (2/60)
2020-12-19 13:46:53,240 [salt.transport.zeromq:1363][DEBUG   ][12596] SaltReqTimeoutError, retrying. (3/60)
```

Go Online

Perform  test.version from the master

Find in the log:

```
2020-12-19 16:24:59,471 [salt.minion      :1671][INFO    ][14724] User root Executing command test.version with jid 20201219172043985731
```

Go Offline again

Wait up to 10 minutes

Find in the log:
- auth_timeout: 60
- auth_tries: 3 **wrong should be 60**

```
2020-12-19 15:18:31,059 [salt.transport.zeromq:1363][DEBUG   ][12596] SaltReqTimeoutError, retrying. (1/3)
2020-12-19 15:19:31,063 [salt.transport.zeromq:1363][DEBUG   ][12596] SaltReqTimeoutError, retrying. (2/3)
2020-12-19 15:20:31,075 [salt.transport.zeromq:1363][DEBUG   ][12596] SaltReqTimeoutError, retrying. (3/3)
```

**Expected behavior**

``` 
2020-12-19 15:18:31,059 [salt.transport.zeromq:1363][DEBUG   ][12596] SaltReqTimeoutError, retrying. (1/60)
```

**Versions Report**
<details><summary>salt --versions-report</summary> 

```
Salt Version:
          Salt: 3002.2

Dependency Versions:
          cffi: 1.12.2
      cherrypy: 17.4.1
      dateutil: 2.8.0
     docker-py: Not Installed
         gitdb: 2.0.5
     gitpython: 2.1.10
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: 1.0.7
       msgpack: 1.0.0
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.19
      pycrypto: Not Installed
  pycryptodome: 3.9.8
        pygit2: Not Installed
        Python: 3.7.4 (tags/v3.7.4:e09359112e, Jul  8 2019, 20:34:20) [MSC v.1916 64 bit (AMD64)]
  python-gnupg: 0.4.4
        PyYAML: 5.3.1
         PyZMQ: 18.0.1
         smmap: 2.0.5
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.1

System Versions:
          dist:
        locale: cp1252
       machine: AMD64
       release: 10
        system: Windows
       version: 10 10.0.18362 SP0

```
</details>

**Additional context**

The consecutive of the log is related to issue https://github.com/saltstack/salt/issues/59064

```
2020-12-19 15:21:31,085 [salt.transport.zeromq:1330][DEBUG   ][12596] Re-init ZMQ socket: Message timed out
2020-12-19 15:21:31,085 [salt.transport.zeromq:1302][DEBUG   ][12596] Trying to connect to: tcp://2.148.96.188:4506
2020-12-19 15:21:31,085 [salt.transport.zeromq:291 ][DEBUG   ][12596] Closing AsyncZeroMQReqChannel instance
2020-12-19 15:21:31,085 [tornado.application:640 ][ERROR   ][12596] Exception in callback functools.partial(<function wrap.<locals>.wrapped at 0x0000011D54AA7B88>, <salt.ext.tornado.concurrent.Future object at 0x0000011D376F2848>)
Traceback (most recent call last):
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\ioloop.py", line 606, in _run_callback
    ret = callback()
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\stack_context.py", line 346, in wrapped
    raise_exc_info(exc)
  File "<string>", line 4, in raise_exc_info
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\stack_context.py", line 317, in wrapped
    ret = fn(*args, **kwargs)
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\gen.py", line 291, in <lambda>
    future, lambda future: callback(future.result()))
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\concurrent.py", line 249, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\gen.py", line 1064, in run
    yielded = self.gen.throw(*exc_info)
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\minion.py", line 1587, in _send_req_async
    ret = yield channel.send(load, timeout=timeout)
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\gen.py", line 1056, in run
    value = future.result()
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\concurrent.py", line 249, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\gen.py", line 1064, in run
    yielded = self.gen.throw(*exc_info)
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\transport\zeromq.py", line 447, in send
    load, tries=tries, timeout=timeout, raw=raw
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\gen.py", line 1056, in run
    value = future.result()
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\concurrent.py", line 249, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\gen.py", line 1064, in run
    yielded = self.gen.throw(*exc_info)
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\transport\zeromq.py", line 416, in _crypted_transfer
    ret = yield _do_transfer()
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\gen.py", line 1056, in run
    value = future.result()
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\concurrent.py", line 249, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\gen.py", line 1064, in run
    yielded = self.gen.throw(*exc_info)
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\transport\zeromq.py", line 399, in _do_transfer
    tries=tries,
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\gen.py", line 1056, in run
    value = future.result()
  File "c:\salt\bin\lib\site-packages\salt-3002.2-py3.7.egg\salt\ext\tornado\concurrent.py", line 249, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
salt.exceptions.SaltReqTimeoutError: Message timed out
```

### PR DESCRIPTION
**Pull request by marbx**
### What does this PR do?
There was code that was overriding the value of `auth_tries` regardless
of what was set in the configuration. It was overriding it to be 3.

### What issues does this PR fix or reference?
This change:
- has been tested to fix #59174 according to its test-procedure
- can serve as a workaround for issue #59064 (Minion throws SaltReqTimeoutError when offline)

### Previous Behavior
The send function never tried to read `auth_tries` from configuration or another source.

### New Behavior
Because the send function does read `detect_mode` from configuration, a successful attempt was made to also read `auth_tries` from configuration. 

### Merge requirements satisfied?

This Draft PR could be used as a hint or even serve as starting point for a Non-Draft Pull request.

**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/transport/zeromq.py [hunk 1]
```diff
@@ -1367,11 +1367,14 @@ def timeout_message(self, message):
...
 
     def send(
-        self, message, timeout=None, tries=3, future=None, callback=None, raw=False
+        self, message, timeout=None, tries=None, future=None, callback=None, raw=False
     ):
         """
         Return a future which will be completed when the message has a response
         """
+        # WIP Retrieve auth_tries from configuration before using 3
+        tries = self.opts.get("auth_tries", 3)
```
Comment (twangboy) [inline salt/transport/zeromq.py:1376]:
```
if tries is None:
    tries = self.opts.get("auth_tries", 3)
```

**Review by twangboy — CHANGES_REQUESTED** [review]

File: salt/transport/zeromq.py [hunk 2]
```diff
@@ -1373,11 +1373,16 @@ def timeout_message(self, message):
...
-        self, message, timeout=None, tries=3, future=None, callback=None, raw=False
+        self, message, timeout=None, tries=None, future=None, callback=None, raw=False
     ):
         """
         Return a future which will be completed when the message has a response
         """
+        # Retrieve auth_tries from configuration if None is passed
+        # Otherwise, use 3
+        if tries is None:
+            tries = self.opts.get("auth_tries", 3)
```
Comment (s0undt3ch) [inline salt/transport/zeromq.py:1384]:
```suggestion
            tries = self.opts.get("auth_tries") or 3
```
If `auth_tries` is set to None, we would never default to 3.

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
We should also add a test where we make sure different `auth_tries` in the config are not ignored

**Comment by twangboy** [comment]
Re-run all

**Review by s0undt3ch — APPROVED** [review]

**Review by Ch3LL — CHANGES_REQUESTED** [review]
>"There was code that was overriding the value of auth_tries regardless"

What part of the code was overriding the `auth_tries`?

I would assume the tcp transport would also be impacted by this and this change would not resolve that transport as well. Thats why I'm trying to understand where we are overriding `auth_tries` because I'm wondering if the fix should be higher in the stack. 

This change would also require test coverage and a changelog file

**Comment by twangboy** [comment]
I don't know how to write tests for this... and I don't know where in the stack to fix this.

**Comment by marbx** [comment]
> > "There was code that was overriding the value of auth_tries regardless"
> 
> What part of the code was overriding the `auth_tries`?
> 
It was the default value `3` of the parameter `tries` in function `send`, which is now replaced by `None`

> I would assume the tcp transport would also be impacted by this and this change would not resolve that transport as well. Thats why I'm trying to understand where we are overriding `auth_tries` because I'm wondering if the fix should be higher in the stack.
> 
> This change would also require test coverage and a changelog file

In the stacktrace of https://github.com/saltstack/salt/issues/59064, I find the zmqtransport , but not the TCP transport. Where do you find the TCP transport?

I would like to better understand how Salt uses ZMQ, where should I start?

**Comment by marbx** [comment]
> I don't know how to write tests for this... and I don't know where in the stack to fix this.

@twangboy: who can?

**Comment by Ch3LL** [comment]
I did a little more investigation here. Upon investigation of your stack trace I noticed the call that is sending `3` is here: 

```
  File "C:\salt\bin\lib\site-packages\salt-3002.6-py3.7.egg\salt\minion.py", line 1587, in _send_req_async
    ret = yield channel.send(load, timeout=timeout)
```

Looks like this PR https://github.com/saltstack/salt/pull/59237 added a new option: `return_retry_tries` which sets it at 3. 

If you increase that option to 5 in your configuration does it resolve the issue for you?

**Comment by Ch3LL** [comment]
err i mean set it to whatever value you want. I think in your original issue you were trying to set it to 60.

**Comment by marbx** [comment]
@Ch3LL , cool finding😃

I only have a smartphone, now.

By looking at the code, #59237seems to fix this issue, but doesnt `return_retry_tries` duplicate `auth_tries`?


`auth_tries` seems to be obsolete.

**Comment by Ch3LL** [comment]
`auth_tries` and `return_retry_tries` accomplish two different tasks. From the docs:

`auth_tries`:
> The number of attempts to authenticate to a master before giving up. Or, more technically, the number of consecutive SaltReqTimeoutErrors that are acceptable when trying to authenticate to the master.

`return_retry_tries`:
> The maximum number of retries for a minion return attempt.

So `auth_tries` is not obsolete but is used for when the minion is attempting to authenticate.

**Comment by Ch3LL** [comment]
Based off of this information are you okay closing this PR?

**Comment by marbx** [comment]
I want to test that

**Comment by Ch3LL** [comment]
@marbx  any luck testing this?

**Comment by marbx** [comment]
I tested above setting, the test failed, unfortunately

**Comment by marbx** [comment]
Please provide an installer or instructions how to test the release which contains `return_retry_tries`, or perform the test, ideally automatically.

**Comment by Ch3LL** [comment]
The change to add that configuration is in all of these releases:

```
v3003
v3003.0rc0
v3003.1
v3003.2
v3003_docs
v3003rc0
v3003rc1
```

So you can install the latest v3003.2 in all of the similar ways you can install Salt today: packages, pip, or installing from source.

**Pull request closed without merging**